### ISSUE TITLE
Results using ST-P3 and VAD metrics

### ISSUE THREAD

**Issue by wljungbergh** [OpenDriveLab/UniAD#166]
Hi, 

As mentioned [here](https://github.com/OpenDriveLab/UniAD/issues/29#issuecomment-1717594344) you are not using the same definition of the L2 metric as other works (e.g., [VAD](https://github.com/hustvl/VAD) or [ST-P3](https://github.com/OpenDriveLab/ST-P3)). 

However, you present UniADs numbers using your new metric against ST-P3s numbers using the old metric. This caused us (and surely many others) to assume that UniAD uses the same definition as in ST-P3 (for example VAD makes this assumption when comparing to your method - see [here](https://github.com/hustvl/VAD/issues/47)).

This greatly underrepresents UniADs performance, unfortunately.  To reduce confusion, it would greatly help if you present your numbers using both definitions.

We ran the numbers for you, and not so surprisingly, your scores improved a lot. Especially at the longer horizons. We include VAD's numbers to highlight that the definition makes a big difference when comparing methods against each other.

Here are your displacement values when using both  metric definitions:
| Method | L2 (m) 1s | L2 (m) 2s | L2 (m) 3s | 
| :---: | :---: | :---: | :---: | 
| ST-P3 | 1.33 | 2.11 | 2.90 |
| UniAD (your metric) | 0.48 | 0.96 | 1.65 | 
| UniAD (old ST-P3 metric) | 0.42 | **0.64** | **0.91** | 
| VAD-Tiny | 0.46 | 0.76 | 1.12 | 
| VAD-Base | **0.41** | 0.70 | 1.05 | 

FYI,  we simply changed the code from
```
for i in range(len(value)):
    row_value.append("%.4f" % float(value[i]))
```
 to 
```
for i in range(len(value)):
    row_value.append("%.4f" % float(value[:i+1].mean()))
```

**Comment by wljungbergh** [comment]
No worries, glad to help. I'd be happy to make a pull request, but I will take up this again sometime after March 7th for obvious reasons. :writing_hand: :page_facing_up:

**Comment by YTEP-ZHI** [comment]
Got it. Good luck!!!---- Replied Message ----FromWilliam ***@***.***>Date03/01/2024 16:02 ***@***.***> CcJiazhi ***@***.***>***@***.***>SubjectRe: [OpenDriveLab/UniAD] Results using ST-P3 and VAD metrics (Issue #166)
No worries, glad to help. I'd be happy to make a pull request, but I will take up this again sometime after March 7th for obvious reasons. ✍️ 

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you commented.Message ID: ***@***.***>
[
{
***@***.***": "http://schema.org",
***@***.***": "EmailMessage",
"potentialAction": {
***@***.***": "ViewAction",
"target": "https://github.com/OpenDriveLab/UniAD/issues/166#issuecomment-1972702523",
"url": "https://github.com/OpenDriveLab/UniAD/issues/166#issuecomment-1972702523",
"name": "View Issue"
},
"description": "View this Issue on GitHub",
"publisher": {
***@***.***": "Organization",
"name": "GitHub",
"url": "https://github.com"
}
}
]

**Comment by wljungbergh** [comment]
Alright! Now I have some time on my hands. Do you want to change the evaluation to match those of SPT3 and VAD? Or do you want it to be configurable? 

Also, where would be the most appropriate place to present the results? The TRAIN_EVAL.md file or in the main README?

**Comment by YTEP-ZHI** [comment]
Hi! I prefer it to be configurable. How about distinguishing the two evaluation protocols by the paper name? One can specify the evaluation strategy using `evaluation_strategy='stp3' or 'uniad'` in their config, and the default value is `uniad`.

**Comment by YTEP-ZHI** [comment]
I will update the README and other instruction files accordingly. No need to worry about it.

**Comment by YTEP-ZHI** [comment]
Also, I want to know whether the misalignment issue happens on the `collision rate` metric? Is the stp3 metric cumulative while uniad is non-cumulative? Have you looked into that before?

**Comment by wljungbergh** [comment]
Yes. They do the same for L2 and collision-rate (i.e., cumulative). As you can see [here](https://github.com/hustvl/VAD/blob/1ede800a517653837faf78b5ec1c3b5fc3ec77b5/projects/mmdet3d_plugin/VAD/VAD.py#L624C1-L629C1), they create one planning metric for each of the timesteps they are using and then take the mean over the values they have added (see [here](https://github.com/hustvl/VAD/blob/1ede800a517653837faf78b5ec1c3b5fc3ec77b5/projects/mmdet3d_plugin/VAD/VAD.py#L634C1-L635C90)). You do this non-cumulative, just as with the L2 definition. 

And yes, if we change that definition too, UniAD achives lower CR than other works.

**Comment by YTEP-ZHI** [comment]
Understood. After modifying the code, could you please make a complete table for both L2 and CR metrics under the `stp3` and `uniad` strategies like the one you posted? 
<img width="468" alt="image" src="https://github.com/OpenDriveLab/UniAD/assets/48089846/5c526a00-a956-4a6f-905b-8a0c2c4c0c02">

It would be of great help if the table included L2 and CR results for STP3, UniAD, and VAD under the two different evaluation strategies.

**Comment by wljungbergh** [comment]
Added a PR #173. Cross-posting the table here:

| Method | L2 1s (m) | L2 2s (m) | L2 3s (m) | CR 1s (%) | CR 2s (%) | CR 3s (%) |
| :---: | :---: | :---: | :---: | :---: | :---: | :---: |
| ST-P3 | 1.33 | 2.11 | 2.90 | 0.23 | 0.62 | 1.27 |
| UniAD (your current metric) | 0.51 | 0.98 | 1.65 | 0.10 | 0.15 | 0.61 |
| UniAD (legacy ST-P3 metric) | 0.42 | **0.64** | **0.91** | **0.07** | **0.10** | **0.22** |
| VAD-Tiny | 0.46 | 0.76 | 1.12 | 0.21 | 0.35 | 0.58 |
| VAD-Base | **0.41** | 0.70 | 1.05 | **0.07** | 0.17 | 0.41 |

**Comment by YTEP-ZHI** [comment]
Great! Thanks for your help.
